numpy/core/src/multiarray/descriptor.c

```diff
@@ -3356,42 +3356,37 @@ arraydescr_struct_str(PyArray_Descr *dtype, int includealignflag)
         sub = arraydescr_struct_dict_str(dtype, includealignflag);
     }
 
-    /* If the data type has a non-void (subclassed) type, show it */
-    if (dtype->type_num == NPY_VOID && dtype->typeobj != &PyVoidArrType_Type) {
+    /* If the data type isn't the default, void, show it */
+    if (dtype->typeobj != &PyVoidArrType_Type) {
         /*
          * Note: We cannot get the type name from dtype->typeobj->tp_name
          * because its value depends on whether the type is dynamically or
          * statically allocated.  Instead use __name__ and __module__.
          * See https://docs.python.org/2/c-api/typeobj.html.
          */
 
-        PyObject *str_name, *namestr, *str_module, *modulestr, *ret;
+        PyObject *namestr, *modulestr, *fmt, *ret;
 
-        str_name = PyUString_FromString("__name__");
-        namestr = PyObject_GetAttr((PyObject*)(dtype->typeobj), str_name);
-        Py_DECREF(str_name);
+        namestr = PyObject_GetAttrString((PyObject *)dtype->typeobj, "__name__");
+        modulestr = PyObject_GetAttrString((PyObject *)dtype->typeobj, "__module__");
 
-        if (namestr == NULL) {
-            /* this should never happen since types always have __name__ */
+        if (namestr == NULL || modulestr == NULL) {
+            /* this should never happen since types always have these attributes */
             PyErr_Format(PyExc_RuntimeError,
-                         "dtype does not have a __name__ attribute");
+                         "dtype.type is missing a __name__ or __module__ "
+                         "attribute");
+            Py_XDECREF(namestr);
+            Py_XDECREF(modulestr);
             return NULL;
         }
 
-        str_module = PyUString_FromString("__module__");
-        modulestr = PyObject_GetAttr((PyObject*)(dtype->typeobj), str_module);
-        Py_DECREF(str_module);
 
-        ret = PyUString_FromString("(");
-        if (modulestr != NULL) {
-            /* Note: if modulestr == NULL, the type is unpicklable */
```

Comment:
More background behind this comment would be useful. Can this ever happen?

Reply:
Hmm my memory of this is vague now. I think this code had to do with "old style" vs "new style" classes in python2. I think "old style" classes fill in the "__module__" entry incompletely, or something along those lines. I will have to look it up.

Reply:
Ah, happily I left some kind of explanation of that in a github comment. See [here (link)](https://github.com/numpy/numpy/pull/5483#issuecomment-71233980).

Now as to whether this can be NULL, I think it can, for types defined in C in python 2. If you go to https://docs.python.org/2/extending/newtypes.html, and copy the first "noddy" example, you will find:
```python
>>> import noddy
>>> mynoddy = noddy.Noddy()
>>> mynoddy.__module__
AttributeError: 'noddy.Noddy' object has no attribute '__module__'
>>> import cPickle
>>>  cPickle.dumps(mynoddy)
TypeError: can't pickle Noddy objects
```

Reply:
So, should the code continue to be able to deal with `__module__ == NULL`? Or is the new behaviour of raising an exception fine? It would probably be good to add a comment with the conclusion regardless...

Reply:
It's probably fine to raise an exception if it is NULL, let's go with these changes.I would change the comment, though, it is technically wrong: Instead of ` this should never happen since types always have these attributes` it should say something like `we do not expect this to happen happen since types always have these attributes, except in some very special cases (module may be null for "statically allocated" types in python2)`.

Probably these "special cases" never come up in practice, since they will cause people other problems (eg, unpicklable types).

Reply:
One last comment: I looked more closely at this and I have not been able to construct an example where `__module__` is NULL.  The type `from numpy.core.test_rational import rational` is what I expected to fail, since it does not properly define a module, but if you do `rational.__module__` you get `__builtin__`, which is not NULL. I tested that `np.dtype((rational, 'i4,i4'))` works properly with this PR.

So in summary, this code is fine.